File: dsa_leetcode_beginner/0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        vector<int> ioList;
        helper(root, ioList);

        bool isBST = true;
        int prev = ioList[0];

        for (int i = 1; i < ioList.size(); i++) {
            if (ioList[i] <= prev) {
                isBST = false;
            }
            prev = ioList[i];
        }

        return isBST;
    }

private:
    void helper(TreeNode* node, vector<int>& ioList) {
        if (node == nullptr) return;

        helper(node->left, ioList);
        ioList.push_back(node->val);
        helper(node->right, ioList);
    }
};

File: dsa_leetcode_beginner/0098-validate-binary-search-tree/0098-validate-binary-search-tree.js
class TreeNode {
    constructor(val = 0, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    isValidBST(root) {
        let ioList = [];
        this.helper(root, ioList);

        let isBST = true;
        let prev = ioList[0];

        for (let i = 1; i < ioList.length; i++) {
            if (ioList[i] <= prev) {
                isBST = false;
            }
            prev = ioList[i];
        }

        return isBST;
    }

    helper(node, ioList) {
        if (node === null) return;

        this.helper(node.left, ioList);
        ioList.push(node.val);
        this.helper(node.right, ioList);
    }
}
